zcash_client_backend/src/data_api/chain.rs

```diff
@@ -170,61 +172,73 @@ where
             .map(|(tag, ivk)| (tag, PreparedIncomingViewingKey::new(&ivk))),
     );
 
-    // Start at either the provided height, or where we synced up to previously.
-    let (scan_from, mut prior_block_metadata) = match from_height {
-        Some(h) => {
-            // if we are provided with a starting height, obtain the metadata for the previous
-            // block (if any is available)
-            (
-                Some(h),
-                if h > BlockHeight::from(0) {
-                    data_db.block_metadata(h - 1).map_err(Error::Wallet)?
-                } else {
-                    None
-                },
-            )
-        }
-        None => {
-            let last_scanned = data_db.block_fully_scanned().map_err(Error::Wallet)?;
-            last_scanned.map_or_else(|| (None, None), |m| (Some(m.block_height + 1), Some(m)))
-        }
+    let mut prior_block_metadata = if from_height > BlockHeight::from(0) {
+        data_db
+            .block_metadata(from_height - 1)
+            .map_err(Error::Wallet)?
+    } else {
+        None
     };
 
-    block_source.with_blocks::<_, DbT::Error>(scan_from, limit, |block: CompactBlock| {
-        add_block_to_runner(params, block, &mut batch_runner);
-        Ok(())
-    })?;
+    let mut continuity_check_metadata = prior_block_metadata;
+    block_source.with_blocks::<_, DbT::Error>(
+        Some(from_height),
+        Some(limit),
+        |block: CompactBlock| {
+            // check block continuity
+            if let Some(scan_error) = check_continuity(&block, continuity_check_metadata.as_ref()) {
+                return Err(Error::Scan(scan_error));
+            }
+            continuity_check_metadata = continuity_check_metadata.as_ref().map(|m| {
+                BlockMetadata::from_parts(
+                    block.height(),
+                    block.hash(),
+                    m.sapling_tree_size()
+                        + u32::try_from(block.vtx.iter().map(|tx| tx.outputs.len()).sum::<usize>())
+                            .unwrap(),
```

Comment:
We iterated over `block.vtx` inside `check_continuity` in order to validate its chain metadata if present, so we should use it if present to avoid duplicate iteration:
```suggestion
                    block.chain_metadata
                        .as_ref()
                        .map(|m| m.sapling_commitment_tree_size)
                        .unwrap_or_else(|| {
                            m.sapling_tree_size()
                                + u32::try_from(block.vtx.iter().map(|tx| tx.outputs.len())
                                    .sum::<usize>())
                                    .unwrap()
                        }),
```